src/legacy/shrink.rs

```diff
@@ -0,0 +1,379 @@
+    };
+
+    debug_assert!(curr_code != CONTROL_CODE as u16);
+    if curr_code > u8::MAX as u16 {
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidData,
+            "the first code must be a literal",
+        ));
+    }
+    let mut first_byte = curr_code as u8;
+    codetab[curr_code as usize].last_dst_pos = dst.len();
+    dst.push_back(curr_code as u8);
+
+    let mut prev_code = curr_code;
+    while dst.len() < uncompressed_size {
+        let Ok(curr_code) = read_code(&mut is, &mut code_size, &mut codetab, &mut queue) else {
+            break;
+        };
+
+        let Some(curr_code) = curr_code else {
+            return Err(Error::new(io::ErrorKind::InvalidData, "invalid code"));
+        };
+
+        let dst_pos = dst.len();
+        // Handle KwKwK: next code used before being added.
+        if Some(curr_code) == queue.next() {
+            if codetab[prev_code as usize].prefix_code.is_none() {
+                return Err(Error::new(
+                    io::ErrorKind::InvalidData,
+                    "previous code no longer valid",
+                ));
+            }
+            // Extend the previous code with its first byte.
+            debug_assert!(curr_code != prev_code);
+            codetab[curr_code as usize].prefix_code = Some(prev_code);
+            codetab[curr_code as usize].ext_byte = first_byte;
+            codetab[curr_code as usize].len = codetab[prev_code as usize].len + 1;
+            codetab[curr_code as usize].last_dst_pos = codetab[prev_code as usize].last_dst_pos;
+            //  dst.push_back(first_byte);
+        }
+
+        // Output the string represented by the current code.
+        let mut len = 0;
+        if let Err(s) = output_code(
+            curr_code,
+            dst,
+            prev_code,
+            &mut codetab,
+            &mut queue,
+            &mut first_byte,
+            &mut len,
+        ) {
+            return Err(s);
+        }
+
+        // Add a new code to the string table if there's room.
+        // The string is the previous code's string extended with
+        // the first byte of the current code's string.
+        let new_code = queue.remove_next();
+        if let Some(new_code) = new_code {
+            //debug_assert!(codetab[prev_code as usize].last_dst_pos < dst_pos);
+            codetab[new_code as usize].prefix_code = Some(prev_code);
+            codetab[new_code as usize].ext_byte = first_byte;
+            codetab[new_code as usize].len = codetab[prev_code as usize].len + 1;
+            codetab[new_code as usize].last_dst_pos = codetab[prev_code as usize].last_dst_pos;
+
+            if codetab[prev_code as usize].prefix_code.is_none() {
+                // prev_code was invalidated in a partial
+                // clearing. Until that code is re-used, the
+                // string represented by new_code is
+                // indeterminate.
+                codetab[new_code as usize].len = UNKNOWN_LEN;
```

Comment:
```suggestion
            let prev_code_entry = codetab[prev_code as usize];
            *codetab[new_code as usize] = Codetab {
                prefix_code: Some(prev_code),
                ext_byte: first_byte,
                last_dst_pos: prev_code_entry.last_dst_pos,
                len: if prev_code_entry.prefix_code.is_none() {
                    // prev_code was invalidated in a partial
                    // clearing. Until that code is re-used, the
                    // string represented by new_code is
                    // indeterminate.
                    UNKNOWN_LEN
                } else {
                    prev_code_entry.len + 1;
                },
            };
```

Reply:
Doesn't look like an improvement -  does that fix something? It's not 1:1 the same.